docs/services/piwik.rst

```diff
@@ -145,11 +145,24 @@ set the context variable ``analytical_identity`` (for global configuration) or
 Disabling cookies
 -----------------
 
-If you want to `disable cookies`_, set :data:`PIWIKI_DISABLE_COOKIES` to
+If you want to `disable cookies`_, set :data:`PIWIK_DISABLE_COOKIES` to
 :const:`True`. This is disabled by default.
 
 .. _`disable cookies`: https://matomo.org/faq/general/faq_157/
 
+Ask for consent
+-----------------
+
+If you want to ask for consent set :data:`PIWIK_ASK_FOR_CONSENT` to
+:const:`True`. This is disabled by default.
+
+To ask the visitor for consent just create DOM elements with the following id's:
+
+`piwik_deny_consent` - id for DOM element to click, if the user denies consent
+`piwik_give_consent` - id for DOM element to click, if the user gives consent
```

Comment:
Yes, Matomo provides an opt-out form. But as far as I understood the concept of "Ask for consent", it is quite independent to the idea of the opt-out form.

With the opt-out form you are tracked until you opt-out (obviously).
With asking for consent, you are not tracked until you deliberately accept the tracking. Therefore, a user/visitor interaction is needed. 

In my opinion, this is the easiest way to implement that feature. The user of django-analytical just has to insert two buttons to his consent request box with these id's.

Reply:
What I'm worried about is that this may be already "too difficult" for a lot of average Django users. You know, developers are spoiled, they want to plug an app in and see it working. :see_no_evil: :grimacing: 

Seriously, the buttons are not the problem, the styling and overall user experience is. ("How should this thing of 'consent or no' look like?", "Where do I place the buttons?", "Do I make them disappear once they are clicked?", etc.)

Don't you think this is challenging for some developers? Or they'd come up with mediocre results (and people start re-iterating the "Wordpress is much better" :smirk:, because Wordpress plugin are designed to have a ready-to-use plug-and-play developer experience).

Can we solve this issue? Matomo doesn't provide anything we can just pull in, does it? Or do the Matomo docs provide a UX design example? (I didn't find any.)

Reply:
I can understand your concern regarding usability. Nevertheless, I don't want to constrain the flexibility to implement the consent buttons as the user wants. I thought one of the reasons users choose Django over Wordpress is that they have more flexibility. Correct me if that's not the case ;)

The idea for a template tag is very good. I would implement it with a start and end tag, like this:

`{ piwik_allow }
<button>Track me!</button>
{ end_piwik_allow }`

You may ask: why not just one tag which uses under the hood a HTML button?
For most of the cases buttons might suffice. But sometimes you want to have only a clickable image, or a font-awesome icon. To preserve flexibility the tag should be able to use arbitrary HTML elements inside of it.

`{ piwik_deny }
<i fa-2x fa-cross></i>
{ end_piwik_deny }`

Is this a possible solution or is it then again too complex?

Reply:
I think the latter is a particularly good idea! And, yes, it's certainly possible (take a look at the [built-in template tags](https://docs.djangoproject.com/en/stable/ref/templates/builtins/) for reference).

The [simple_tag](https://docs.djangoproject.com/en/stable/howto/custom-template-tags/#writing-custom-template-tags) can have arguments, so our API may look like this:

```django
{ piwik_allow }  yields just  id="..."  (as in your example)
```
```django
{ piwik_allow button }  yields  <button id="...">Allow</button>  (localized)
```
```django
{ piwik_allow button "Sure!" }  yields  <button id="...">Sure!</button>
```
```django
{ piwik_deny div }  yields  <div id="...">Deny</div>  (just like <button>)
```
```django
{ piwik_deny i "fa-2x fa-cross" }  yields  <i id="..." fa-2x fa-cross></i>
```
Just allow a few ones. Depending on the to-be-generated HTML tag you validate and use arguments 2+ the appropriate way (e.g. use parameter 3 for `class="..."` or so, if that's really needed).

That looks sufficiently clean, doesn't it? How do you like it? :smiley:

Reply:
I have implemented it like you proposed, but now I have doubts.

I like the idea but it's maybe problematic that sometimes the template tag returns fully fledged HTML and sometimes it returns just an id.

Also, there are many edge cases, like what if a user wants to set the "type" attribute of an `input` element, or the user leaves out some parameters. This means I have to enhance the template tag further, and validate the input, etc.

Furthermore, now the user writes
`{% piwik_remember_consent "button" "Please track me!" "buttonClass" %}`
instead of 
`<button {% piwik_remember_consent %} class="buttonClass">Track me!</button>`

In my opinion, this is even more difficult to use, in case you don't know template tags very well.

I would go with a simple template tag: the user builds his own HTML and attaches the id with the tag. This would keep the template tag function slim and coherent and gives the whole flexibility to the user, while we do not have to worry about edge cases.

What do you think?

Reply:
Sure, the usage you prefer would still be the default. Hence, I wouldn't worry.

The "button" & friends, on the other hand, I see as nice idioms: You say, _"Don't give me just _something_, which helps me to construct a consent UI control: Give me a button!"_

Note that I wouldn't use quotes for writing _"button"_, e.g.
```django
{% piwik_remember_consent button %}
```
You have the full freedom with the simple tag and get a nice, of course modified but idiomatic, result with additional parameters.

Let's get another opinion on board, though: @jcassee, what do you think?

Reply:
Don't worry too much about edge cases, btw. It's fine when the short idiomatic solutions only cover the principal use cases.

Reply:
First of all: I'm not sure that pushing the event to Piwik is the _only_ thing the consent buttons need to do. They probably also need to close a dialog or fixed banner. Depending on the JS tech used, you need the tag id to integrate with other interactions. This make me think that using classes is better than using the ids.

Looking over the different options, the original one still seems simplest (using a class):
```html
<button class="big green piwik_give_consent">I agree!</button>
```

Using a Django tag that "hides" the HTML tag makes styling with CSS a bit weirder and is not easily customizable without abstracting HTML attributes inside the Django tag (`{ piwik_deny i "fa-2x fa-cross" }`). Wrapping an HTML tag with a Django tag (`{ piwik_deny } <i fa-2x fa-cross></i> { end_piwik_deny }`) looks pretty complex.

Reply:
I implemented it now as @jcassee proposed because I think it makes most sense in this way and doesn't get to convoluted. Please check the new version and let me know if there is something missing to get this pull request merged.

Thanks for the fast and nice collaboration, I really enjoyed my first PR :)